src/workerd/api/streams/standard.c++

```diff
@@ -3032,41 +3057,47 @@ private:
+                    // possible for the promise that is holding the PumpToReader to be
+                    // dropped causing the hold on the sink to be released. If that is
+                    // released while the write is still pending we can end up with an
+                    // error further up the destruct chain.
+                    return ioContext.awaitIo(js, reader->canceler.wrap(kj::mv(promise)),
+                        [](jsg::Lock& js) -> kj::Maybe<jsg::Value> {
+                      // The write completed successfully.
+                      return kj::Maybe<jsg::Value>(nullptr);
+                    }, [](jsg::Lock& js, jsg::Value exception) mutable -> kj::Maybe<jsg::Value> {
+                      // The write failed.
+                      return kj::mv(exception);
+                    }).then(js, ioContext.addFunctor(
+                        [readable=kj::mv(readable),pumpToReader=kj::mv(pumpToReader)]
+                        (jsg::Lock& js, kj::Maybe<jsg::Value> maybeException) mutable {
+                      KJ_IF_MAYBE(reader, tryGetAs<PumpToReader>(pumpToReader)) {
+                        auto& ioContext = reader->ioContext;
+                        ioContext.requireCurrentOrThrowJs();
+                        // Oh good, if we got here it means we're in the right IoContext and
+                        // the PumpToReader is still alive.
+                        KJ_IF_MAYBE(exception, maybeException) {
+                          reader->doError(js, exception->getHandle(js));
+                        }
+                        return reader->pumpLoop(js, ioContext, kj::mv(readable), kj::mv(pumpToReader));
+                      } else {
+                        // If we got here, we're in the right IoContext but the PumpToReader
+                        // has been destroyed. Let's cancel the readable as the last step.
+                        return readable->cancel(js, maybeException.map([&](jsg::Value& ex) {
+                          return ex.getHandle(js);
+                        }));
                       }
-                      return reader->pumpLoop(js, ioContext, kj::mv(readable), kj::mv(pumpToReader));
-                    } else {
-                      // If we got here, we're in the right IoContext but the PumpToReader
-                      // has been destroyed. Let's cancel the readable as the last step.
-                      return readable->cancel(js, maybeException.map([&](jsg::Value& ex) {
-                        return ex.getHandle(js);
-                      }));
-                    }
-                  }));
+                    }));
+                  }, [&](jsg::Value exception) {
+                    // Exceptions here should be rare, and fairly odd. We would get here,
+                    // for instance, if the call to .Then inside read().then(...) failed.
+                    return handleException(kj::mv(exception));
```

Comment:
Catching the exception and turning it into a rejected promise seems redundant here, since we're already in a `.then()` callback. If we just let the exception throw out of the callback, the promise machinery will already capture it and reject the promise.